paddle/framework/tensor.cc

```diff
@@ -0,0 +1,95 @@
+#include "paddle/framework/tensor.h"
+
+namespace paddle {
+namespace framework {
+
+template <typename T>
+const T* Tensor::Data() const {
+  PADDLE_ASSERT(holder_ != nullptr);
+  PADDLE_ASSERT(holder_->Place() == place_);
+  PADDLE_ASSERT(holder_->Size() >= product(dims_) * sizeof(T));
+  return static_cast<const T*>(holder->Ptr());
+}
+
+bool Tensor::NeedReset() {
+  return (holder_ == nullptr || holder_->Place() != place_ ||
+          holder_->Size() < product(dims_) * sizeof(T));
```

Comment:
There is no `T`. Do there need a `template`? And again, do we need to check the type of `holder_` here?